cmd/egress-v6-cni-plugin/cni.go

```diff
@@ -0,0 +1,608 @@
+	if err != nil {
+		log.Debugf("enableHostIPv6Forwarding failed: %v", err)
+		return err
+	}
+	log.Debugf("enable host IPv6 forwarding successfully")
+
+	err = setupHostIPv6Route(netns, args.IfName, containerIPNetV6.IP)
+	if err != nil {
+		log.Debugf("setupHostIPv6Route failed: %v", err)
+		return err
+	}
+	log.Debugf("host IPv6 route set up successfully")
+
+	err = setupHostIPv6Snat(containerIPNetV6.IP, hostPrimaryIfIPv6, chain, comment, netConf.RandomizeSNAT)
+	if err != nil {
+		log.Debugf("setupHostIPv6Snat failed: %v", err)
+		return err
+	}
+
+	log.Debugf("host IPv6 SNAT set up successfully")
+
+	// Copy IPs over to result
+	result.IPs = append(result.IPs, ipamResult.IPs...)
+	// Note: Useful for debug, will do away with the below log prior to release
+	for _, v := range result.IPs {
+		log.Debugf("Interface index: %d; IP: %s", *v.Interface, v.Address)
+	}
+
+	// Pass through the previous result
+	return types.PrintResult(result, netConf.CNIVersion)
+}
+
+func cmdDel(args *skel.CmdArgs) error {
+
+	netConf, log, err := loadConf(args.StdinData)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	log.Debugf("Received Del request for: conf=%v; Plugin enabled=%s", netConf, netConf.Enabled)
+
+	// We only need this plugin to kick in if v6 is enabled
+	if netConf.Enabled == "false" {
+		return nil
+	}
+
+	if netConf.PrevResult == nil {
+		log.Debugf("must be called as a chained plugin")
+		return fmt.Errorf("must be called as a chained plugin")
+	}
+
+	result, err := current.GetResult(netConf.PrevResult)
+	if err != nil {
+		log.Debugf("get PrevResult failed: %v", err)
+		return err
+	}
+	log.Debugf("PrevResult: %v", result)
+
+	// figure out container IPv6 address
+	var containerVethIPv6 net.IPNet
+	for _, ipc := range result.IPs {
+		if ipc.Version == "6" && ipc.Address.IP.IsGlobalUnicast() {
+			containerVethIPv6 = ipc.Address
+		}
+	}
+	log.Debugf("container IPv6 address found: %s", containerVethIPv6.String())
+
+	// delete host SNAT - both ip6table nat table's rule and chain for this container's IPv6 traffic
+	chain := utils.MustFormatChainNameWithPrefix(netConf.Name, args.ContainerID, "E6-")
+	comment := utils.FormatComment(netConf.Name, args.ContainerID)
+
+	err = snat.Snat6Del(containerVethIPv6.IP, chain, comment)
+	if err != nil {
+		log.Debugf("Delete host SNAT for container IPv6 %s failed: %v.", containerVethIPv6, err)
+		return nil
+	}
+	log.Debugf("Host ipv6tables snat rule/chain deleted for container")
+
+	netns, err := ns.GetNS(args.Netns)
+	if err != nil {
+		log.Debugf("failed to open netns %q: %v", args.Netns, err)
+		return fmt.Errorf("failed to open netns %q: %v", args.Netns, err)
+	}
+	defer netns.Close()
+
+	// delete host route - return traffic route for container's IPv6 traffic
+	err = tearDownHostIPv6Route(netns, args.IfName, containerVethIPv6.IP)
+	if err != nil {
+		log.Debugf("tear down host IPv6 route for container IPv6: $s failed: %v", containerVethIPv6.String(), err)
+		return err
```

Comment:
CNI deletes should be best effort, so we don't want to return error here. We can only log and continue